statsmodels/gam/tests/test_gam.py

```diff
@@ -0,0 +1,791 @@
+                            [2., 4., 8., 8., 64., 512.]]
+    assert_allclose(train_smoother.basis, expected_train_basis)
+
+    expected_test_basis = [[3., 9., 27., 9., 81., 729.],
+                           [4., 16., 64., 10., 100., 1000.],
+                           [5., 25., 125., 11., 121., 1331.]]
+    assert_allclose(test_smoother.basis, expected_test_basis)
+
+
+def test_get_sqrt():
+    n = 1000
+    np.random.seed(1)
+    x = np.random.normal(0, 1, (n, 3))
+    x2 = np.dot(x.T, x)
+
+    sqrt_x2 = matrix_sqrt(x2)
+
+    x2_reconstruction = np.dot(sqrt_x2.T, sqrt_x2)
+    assert_allclose(x2_reconstruction, x2)
+
+
+def test_make_augmented_matrix():
+    np.random.seed(1)
+    n = 500
+    x = np.random.uniform(-1, 1, (n, 3))
+    s = np.dot(x.T, x)
+    y = np.array(list(range(n)))
+    w = np.random.uniform(0, 1, n)
+    nobs, n_columns = x.shape
+
+    # matrix_sqrt removes redundant rows,
+    # if alpha is zero, then no augmentation is needed
+    alpha = 0
+    aug_y, aug_x, aug_w = make_augmented_matrix(y, x, alpha * s, w)
+    expected_aug_x = x
+    assert_allclose(aug_x, expected_aug_x)
+    expected_aug_y = y
+    expected_aug_y[:nobs] = y
+    assert_allclose(aug_y, expected_aug_y)
+    expected_aug_w = w
+    assert_allclose(aug_w, expected_aug_w)
+
+    alpha = 1
+    aug_y, aug_x, aug_w = make_augmented_matrix(y, x, s, w)
+    rs = matrix_sqrt(alpha * s)
+    # rs = sp.linalg.cholesky(alpha * s)
+    assert_allclose(np.dot(rs.T, rs), alpha * s)
+    expected_aug_x = np.vstack([x, rs])
+    assert_allclose(aug_x, expected_aug_x)
+    expected_aug_y = np.zeros(shape=(nobs + n_columns,))
+    expected_aug_y[:nobs] = y
+    assert_allclose(aug_y, expected_aug_y)
+    expected_aug_w = np.concatenate((w, [1] * n_columns), axis=0)
+    assert_allclose(aug_w, expected_aug_w)
+
+
+def test_penalized_wls():
+    np.random.seed(1)
+    n = 20
+    p = 3
+    x = np.random.normal(0, 1, (n, 3))
+    y = x[:, 1] - x[:, 2] + np.random.normal(0, .1, n)
+    y -= y.mean()
+
+    weights = np.ones(shape=(n,))
+    s = np.random.normal(0, 1, (p, p))
+
+    pen_wls_res = penalized_wls(y, x, 0 * s, weights)
+    ls_res = lm.OLS(y, x).fit()
+
+    assert_allclose(ls_res.params, pen_wls_res.params)
+
+
+def test_cyclic_cubic_splines():
+    cur_dir = os.path.dirname(os.path.abspath(__file__))
+    file_path = os.path.join(cur_dir, "results",
+                             "cubic_cyclic_splines_from_mgcv.csv")
+    data_from_r = pd.read_csv(file_path)
+
+    x = data_from_r[['x0', 'x2']].as_matrix()
+    y = data_from_r['y'].as_matrix()
+    y_est_mgcv = data_from_r[['y_est']].as_matrix()
+    s_mgcv = data_from_r[['s(x0)', 's(x2)']].as_matrix()
+
+    dfs = [10, 10]
+    ccs = CyclicCubicSplines(x, df=dfs)
+    alpha = [0.05 / 2, 0.0005 / 2]
+    # TODO: if alpha changes in pirls this should be updated
+
+    gam = GLMGam(y, smoother=ccs, alpha=alpha)
+    # gam_res = gam._fit_pirls(y, ccs, alpha=alpha)
```

Comment:
?

Reply:
old version, deleted
signature for _fit_pirls has also changed